src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +32,131 @@ public static class Renamer
+
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+
+            var typeDeclarations = syntaxRoot.DescendantNodesAndSelf().Where(syntaxFacts.IsTypeDeclaration);
+
+            var matchingTypeDeclaration = typeDeclarations.FirstOrDefault(d => syntaxFacts.GetDisplayName(d, DisplayNameOptions.None).Equals(oldNameWithoutExtension, StringComparison.OrdinalIgnoreCase));
+            if (matchingTypeDeclaration is object)
+            {
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                solution = await RenameSymbolAsync(solution, symbol, newNameWithoutExtension, optionSet, cancellationToken).ConfigureAwait(false);
+            }
+
+            return solution;
+        }
+
+        private static async Task<Solution> SyncNamespaceToDocumentFoldersAsync(DocumentId documentId, Solution solution, CancellationToken cancellationToken)
+        {
+            var document = solution.GetRequiredDocument(documentId);
+            var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var targetNamespace = BuildNamespaceFromFolders(document.Folders, syntaxFacts);
+
+            var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // Don't descend into anything other than top level declarations from the root.
+            // ChangeNamespaceService only controls top level declarations right now
+            var namespaceDeclarations = syntaxRoot.DescendantNodes(n => !syntaxFacts.IsDeclaration(n)).Where(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+            foreach (var namespaceDeclaration in namespaceDeclarations)
+            {
+                solution = await changeNamespaceService.ChangeNamespaceAsync(
+                    document,
+                    namespaceDeclaration,
+                    targetNamespace,
+                    cancellationToken).ConfigureAwait(false);
+
+                document = solution.GetRequiredDocument(document.Id);
+            }
+
+            return solution;
+
+            static string BuildNamespaceFromFolders(
+                IEnumerable<string> folders,
+                ISyntaxFactsService syntaxFacts)
+            {
+                var parts = folders.SelectMany(folder => folder.Split(new[] { '.' }));
+
+                // string.Empty is used to represent the global namespace
+                // when calling ChangeNamespaceAsync
+                return parts.All(syntaxFacts.IsValidIdentifier) ? string.Join(".", parts) : string.Empty;
```

Comment:
This works if your folders form your entire namespace but what about things like in Roslyn where we have the common root namespace and the folders are suffixes? Or VB default namespaces?